src/librbd/api/Mirror.cc

```diff
@@ -3146,44 +3171,70 @@ int Mirror<I>::group_promote(IoCtx& group_ioctx, const char *group_name,
                  << " :" << cpp_strerror(r) << dendl;
       return r;
     }
-    auto snap =  snaps.rbegin();
+
+    if (snaps.empty()) {
+      lderr(cct) << "cannot rollback, no mirror group snap available on group: "
+                 << group_name << dendl;
+      return -EINVAL;
+    }
+
+    bool need_rollback = false;
+    auto snap = snaps.rbegin();
     for (; snap != snaps.rend(); ++snap) {
-      auto snap_type = cls::rbd::get_group_snap_namespace_type(
-          snap->snapshot_namespace);
-      if (snap_type == cls::rbd::GROUP_SNAPSHOT_NAMESPACE_TYPE_MIRROR &&
-          snap->state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
-        break;
+      auto mirror_ns = std::get_if<cls::rbd::GroupSnapshotNamespaceMirror>(
+          &snap->snapshot_namespace);
+      if (mirror_ns == nullptr) {
+        continue;
+      }
+
+      if (snap->state != cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
+        need_rollback = true;
+        continue;
       }
+      break;
```

Comment:
I'm not sure about this block of code that determines
* whether rollback is feasible
* whether rollback is necessary

As per the code,
* if there is a complete mirror group snap, then rollback is feasible. 
* if the most recent mirror group snap is not complete, then rollback is required.

@nbalacha can you please confirm that the above statements are correct?

If the above statements are true, I think the following code is easier to follow:
```
      auto mirror_ns = std::get_if<cls::rbd::GroupSnapshotNamespaceMirror>(
          &snap->snapshot_namespace);
      if (mirror_ns) {
        if (snap->state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
          // rollback is feasible
          break;
        }
        need_rollback = true;
      }      
```

Reply:
I'm not a fan of nesting thing where ever possible, would leave it as it if its not a big deal.